### PR TITLE
Ensure Logback is refreshed before application starts

### LINKED ISSUE #8139: [GRAALVM] Logback not fetching Environment Variable
### Expected Behavior

We are using [Logstash Logback Encoder](https://github.com/logfellow/logstash-logback-encoder) and we are trying to inject Field Values as Environment Variables.

e.g.:
```
<pattern>
                    <omitEmptyFields>true</omitEmptyFields>
                    <pattern>
                        {
                            "CorrelationId": "%mdc{trace_id}",
                            "Resource": {
                                "application": "demo",
                                "logger": "%logger",
                                "version":"${VERSION}",
                                "environment":"${ENVIRONMENT}",
                            }
                        }
                    </pattern>
                </pattern>
```

### Actual Behaviour

Variables are not properly filled. Result:

```
{
    "Timestamp": "2022-10-10T10:01:16.565Z",
    "SeverityText": "INFO",
    "Resource": {
        "application": "demo",
        "logger": "io.micronaut.function.aws.proxy.AbstractLambdaContainerHandler",
        "version": "VERSION_IS_UNDEFINED"
        "environment": "ENVIRONMENT_IS_UNDEFINED"
    },
    "Body": {
        "message": "Starting Lambda Container Handler",
        "thread": "main"
    }
}

```

### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

https://github.com/driverpt/micronaut-otel-issue

### Version

3.7.1

### PR DESCRIPTION
**Pull request by driverpt**
Attempts to fix #8139

### CODE REVIEW

**Comment by graemerocher** [comment]
I believe this will be fixed by https://github.com/micronaut-projects/micronaut-core/pull/8235 which moves logback to runtime intiailization

**Comment by driverpt** [comment]
@graemerocher , with this modification you can keep build time initialization, it just resets all internal logback caches when Micronaut Inits.

The issue is that Logback is initialized when JVM Starts. The `LoggerContext#reset` only resets it's internal caches, so we can take advantage of Logback being previously initialized at build time. Won't improve much in terms of performance, but will add something

**Comment by graemerocher** [comment]
This seems safe enough to target 3.8.x branch, can you do that?

**Comment by driverpt** [comment]
Done. Can you unblock the CI? I suspect that the test will fail

**Comment by graemerocher** [comment]
not sure why it is blocked 🤔

**Comment by graemerocher** [comment]
@n0tl3ss this might help with the logging issue for AWS/OCI

**Comment by driverpt** [comment]
Is there any ETA on this?

**Comment by graemerocher** [comment]
waiting for feedback from @n0tl3ss

**Comment by driverpt** [comment]
@n0tl3ss , can you help out here? This is currently the only blocker we have to go live with Micronaut

**Comment by graemerocher** [comment]
@driverpt from the looks of the change it seems like you could wire this into your existing application without a new version of Micronaut necessarily blocking you?

**Comment by n0tl3ss** [comment]
Hi, @driverpt. I created copy of this pull request: https://github.com/micronaut-projects/micronaut-core/pull/8263.

**Comment by driverpt** [comment]
@n0tl3ss , fixed checkstyle issue

**Review by n0tl3ss — APPROVED** [review]

**Comment by n0tl3ss** [comment]
> @n0tl3ss this might help with the logging issue for AWS/OCI

I tried but without success

**Comment by graemerocher** [comment]
please fix the broken tests

**Comment by driverpt** [comment]
I will, but at the end of the day

**Comment by driverpt** [comment]
@graemerocher @n0tl3ss , GitHub Workflow needs a maintainers approval to run

File: runtime/src/main/java/io/micronaut/logging/impl/LogbackLoggingSystem.java [hunk 1]
```diff
@@ -35,11 +42,31 @@
...
 
+    @Override
+    public void refresh() {
+        getLoggerContext().reset();
+        JoranConfigurator joranConfigurator = new JoranConfigurator();
+        joranConfigurator.setContext(getLoggerContext());
+        String logbackXml = logbackXmlLocation.orElse(DEFAULT_LOGBACK_LOCATION);
+
+        try {
+            URL resource = getClass().getClassLoader().getResource(logbackXml);
```
Comment (driverpt) [inline runtime/src/main/java/io/micronaut/logging/impl/LogbackLoggingSystem.java]: @graemerocher  @n0tl3ss , please review this part. I need your opinion on this one

Comment (n0tl3ss) [reply runtime/src/main/java/io/micronaut/logging/impl/LogbackLoggingSystem.java]: I think something like
```
    @Override
    public void refresh() {
        getLoggerContext().reset();
        try {
            new ContextInitializer(getLoggerContext()).autoConfig();
        } catch (JoranException e) {
            throw new LoggingSystemException("Error while refreshing Logback", e);
        }
    }
```

 would do a trick.

**Review by n0tl3ss — CHANGES_REQUESTED** [review]
There is `io.micronaut.management.endpoint.loggers.impl.LogbackLoggingSystem` class that might replace `io.micronaut.logging.impl.LogbackLoggingSystem`. Should it also override refresh() method?

**Comment by driverpt** [comment]
> There is `io.micronaut.management.endpoint.loggers.impl.LogbackLoggingSystem` class that might replace `io.micronaut.logging.impl.LogbackLoggingSystem`. Should it also override refresh() method?

I'm not a fan of copy + paste the same code in 2 classes. What do you suggest?

**Comment by driverpt** [comment]
@n0tl3ss , were you able to get this "Fix" running smoothly?

I've been testing this and it seems like after I call `loggerContext.reset()` that it goes completely silent.

**Comment by driverpt** [comment]
@n0tl3ss , can you please test if the logging works in AWS/OCI ?

Thanks

**Comment by driverpt** [comment]
UPDATE: This seems to work in AWS Environment. But the downside is that we will need to add a `reflect-config.json` for all Logback components that we are using.

E.g.: In my case, the `logstash-logback-encoder`. But seems to be working with the simple Console Appender

**Comment by driverpt** [comment]
@graemerocher @n0tl3ss , Checkstyle is failing due to NettyServer.java issue

**Comment by graemerocher** [comment]
Thanks for the contribution!

**Pull request merged**